// @nuxtjs/mcp-toolkit/dist/runtime/server/mcp/deeplink.js
import { defineEventHandler, getRequestURL, getQuery, setHeader } from "h3";
import mcpConfig from "#nuxt-mcp-toolkit/config.mjs";
const IDE_CONFIGS = {
  cursor: {
    name: "Cursor",
    generateDeeplink: (serverName, mcpUrl) => {
      const config = { type: "http", url: mcpUrl };
      const configBase64 = Buffer.from(JSON.stringify(config)).toString("base64");
      return `cursor://anysphere.cursor-deeplink/mcp/install?name=${encodeURIComponent(serverName)}&config=${encodeURIComponent(configBase64)}`;
    }
  },
  vscode: {
    name: "VS Code",
    generateDeeplink: (serverName, mcpUrl) => {
      const config = { name: serverName, type: "http", url: mcpUrl };
      return `vscode:mcp/install?${encodeURIComponent(JSON.stringify(config))}`;
    }
  }
};
function escapeHtmlAttr(str) {
  return str.replace(/&/g, "&amp;").replace(/"/g, "&quot;").replace(/'/g, "&#39;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}
function escapeJs(str) {
  return str.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/'/g, "\\'").replace(/</g, "\\u003c").replace(/>/g, "\\u003e");
}
export default defineEventHandler((event) => {
  const requestUrl = getRequestURL(event);
  const query = getQuery(event);
  const ide = query.ide || "cursor";
  const ideConfig = IDE_CONFIGS[ide];
  if (!ideConfig) {
    setHeader(event, "Location", "/");
    return new Response(null, { status: 302 });
  }
  const serverName = query.name || mcpConfig.name || "mcp-server";
  const mcpUrl = `${requestUrl.origin}${mcpConfig.route || "/mcp"}`;
  const deeplink = ideConfig.generateDeeplink(serverName, mcpUrl);
  const htmlDeeplink = escapeHtmlAttr(deeplink);
  const jsDeeplink = escapeJs(deeplink);
  setHeader(event, "Content-Type", "text/html; charset=utf-8");
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Opening ${ideConfig.name}...</title>
  <style>
    body { font-family: system-ui, sans-serif; display: flex; align-items: center; justify-content: center; min-height: 100vh; margin: 0; background: #0a0a0a; color: #fff; }
    .container { text-align: center; padding: 2rem; }
    a { color: #3b82f6; }
  </style>
</head>
<body>
  <div class="container">
    <p>Opening ${ideConfig.name}...</p>
    <p>If nothing happens, <a href="${htmlDeeplink}">click here to install</a>.</p>
  </div>
  <script>window.location.href = "${jsDeeplink}";<\/script>
</body>
</html>`;
});
