// @nuxtjs/mcp-toolkit/dist/runtime/server/mcp/loaders/index.js
import { addServerTemplate, logger } from "@nuxt/kit";
import {
  createExcludePatterns,
  createTemplateContent,
  findIndexFile,
  loadDefinitionFiles,
  toIdentifier
} from "./utils.js";
const log = logger.withTag("@nuxtjs/mcp-toolkit");
async function loadMcpDefinitions(type, templateFilename, paths) {
  try {
    const result = await loadDefinitionFiles(paths);
    addServerTemplate({
      filename: templateFilename,
      getContents: () => {
        if (result.count === 0) {
          return `export const ${type} = []
`;
        }
        const entries = result.files.map((file) => {
          const filename = file.split("/").pop();
          const identifier = toIdentifier(filename);
          return [identifier, file];
        });
        return createTemplateContent(type, entries);
      }
    });
    return result;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    log.error(`Failed to load ${type} definitions from paths: ${paths.join(", ")}`);
    log.error(`Error: ${errorMessage}`);
    throw new Error(`Failed to load MCP ${type} definitions. Check that the paths exist and contain valid definition files.`);
  }
}
async function loadHandlers(paths = []) {
  try {
    if (paths.length === 0) {
      addServerTemplate({
        filename: "#nuxt-mcp-toolkit/handlers.mjs",
        getContents: () => `export const handlers = []
`
      });
      return { count: 0, files: [], overriddenCount: 0 };
    }
    const excludePatterns = createExcludePatterns(paths, ["tools", "resources", "prompts"]);
    const result = await loadDefinitionFiles(paths, {
      excludePatterns,
      filter: (filePath) => {
        const relativePath = filePath.replace(/.*\/server\//, "");
        const filename = filePath.split("/").pop();
        const isIndexFile = /^index\.(?:ts|js|mts|mjs)$/.test(filename);
        return !relativePath.includes("/tools/") && !relativePath.includes("/resources/") && !relativePath.includes("/prompts/") && !isIndexFile;
      }
    });
    addServerTemplate({
      filename: "#nuxt-mcp-toolkit/handlers.mjs",
      getContents: () => {
        if (result.count === 0) {
          return `export const handlers = []
`;
        }
        const entries = result.files.map((file) => {
          const filename = file.split("/").pop();
          const identifier = toIdentifier(filename);
          return [identifier, file];
        });
        return createTemplateContent("handlers", entries);
      }
    });
    return result;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    log.error(`Failed to load handler definitions from paths: ${paths.join(", ")}`);
    log.error(`Error: ${errorMessage}`);
    throw new Error(`Failed to load MCP handler definitions. Check that the paths exist and contain valid handler files.`);
  }
}
export async function loadTools(paths) {
  return loadMcpDefinitions("tools", "#nuxt-mcp-toolkit/tools.mjs", paths);
}
export async function loadResources(paths) {
  return loadMcpDefinitions("resources", "#nuxt-mcp-toolkit/resources.mjs", paths);
}
export async function loadPrompts(paths) {
  return loadMcpDefinitions("prompts", "#nuxt-mcp-toolkit/prompts.mjs", paths);
}
export { loadHandlers };
async function loadDefaultHandler(paths = []) {
  try {
    const indexFile = await findIndexFile(paths);
    addServerTemplate({
      filename: "#nuxt-mcp-toolkit/default-handler.mjs",
      getContents: () => {
        if (!indexFile) {
          return `export const defaultHandler = null
`;
        }
        return `import handler from '${indexFile}'
export const defaultHandler = handler
`;
      }
    });
    return indexFile !== null;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    log.error(`Failed to load default handler: ${errorMessage}`);
    addServerTemplate({
      filename: "#nuxt-mcp-toolkit/default-handler.mjs",
      getContents: () => `export const defaultHandler = null
`
    });
    return false;
  }
}
export async function loadAllDefinitions(paths) {
  try {
    const [tools, resources, prompts, handlers, hasDefaultHandler] = await Promise.all([
      loadTools(paths.tools),
      loadResources(paths.resources),
      loadPrompts(paths.prompts),
      loadHandlers(paths.handlers ?? []),
      loadDefaultHandler(paths.handlers ?? [])
    ]);
    const results = {
      tools,
      resources,
      prompts,
      handlers,
      hasDefaultHandler
    };
    return {
      ...results,
      total: tools.count + resources.count + prompts.count + handlers.count
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    log.error("Failed to load MCP definitions");
    log.error(`Error: ${errorMessage}`);
    throw new Error("Failed to load MCP definitions. Please check your configuration and ensure definition files are valid.");
  }
}
export async function getHandlerRoutes() {
  try {
    const handlersModule = await import("#nuxt-mcp-toolkit/handlers.mjs");
    const handlers = handlersModule.handlers;
    return handlers.map((h) => ({ name: h.name, route: h.route }));
  } catch {
    return [];
  }
}
