// @prisma/nuxt/dist/module.mjs
import { defineNuxtModule, createResolver, addPlugin, addImportsDir, addServerImportsDir } from '@nuxt/kit';
import { addCustomTab } from '@nuxt/devtools-kit';
import { fileURLToPath } from 'url';
import defu from 'defu';
import { x } from 'tinyexec';
import c from 'tinyrainbow';
import { existsSync, mkdirSync, writeFileSync, readFileSync } from 'fs';
import { join } from 'pathe';
import consola, { consola as consola$1 } from 'consola';

const PREDEFINED_LOG_MESSAGES = {
  checkIfPrismaSchemaExists: {
    yes: "Prisma schema file exists.",
    no: "Prisma schema file does not exist."
  },
  initPrisma: {
    action: "Initializing Prisma project...",
    success: "Initialized the Prisma project.",
    error: "Failed to initialize Prisma project."
  },
  checkIfMigrationsFolderExists: {
    success: "Database migrations folder exists.",
    error: "Database migrations folder does not exist."
  },
  writeToSchema: {
    errorReadingFile: "Error reading existing schema file.",
    failedToWrite: "Failed to write models to Prisma schema."
  },
  runMigration: {
    action: "Migrating database schema...",
    success: "Created User and Post tables in the database.",
    error: "Failed to run Prisma migration."
  },
  formatSchema: {
    action: "Formatting Prisma schema...",
    success: "Successfully formatted Prisma schema.",
    error: "Failed to format Prisma schema file."
  },
  generatePrismaClient: {
    action: "Generating Prisma client...",
    success: "Successfully generated Prisma client.",
    error: "Failed to generate Prisma Client."
  },
  startPrismaStudio: {
    action: "Starting Prisma Studio...",
    success: "Prisma Studio started.",
    info: `After clicking ${c.bold("Get Started")} in Nuxt DevTools, click on the ${c.bold("three dots (\uFE19)")} in the lower left-hand side to reveal additional tabs. Locate the Prisma logo to open Prisma Studio.
`,
    error: "Failed to start Prisma Studio."
  },
  writeClientInLib: {
    found: "Skipping the creation of a lib/prisma.ts file that would hold a global instance of the Prisma Client because the prisma.ts file already exists in the lib folder.",
    success: "Global instance of Prisma Client created in lib/prisma.ts."
  },
  PRISMA_SETUP_SKIPPED_WARNING: "Nuxt Prisma Module setup skipped.\nThis may cause unexpected behavior.",
  skipMigrations: `
Not migrating the database.`,
  skipInstallingPrismaStudio: "Skipped installing Prisma Studio.",
  suggestions: {
    migrate: c.yellow(c.bold("\nHint: ")) + `You can manually run migrations by executing ${c.cyan(c.bold("npx prisma migrate dev"))} or visit the ${c.blue(c.bold("Prisma Migrate"))} docs for more info:
${c.underline(c.blue("https://pris.ly/nuxt/migrate"))}. Or if you have pre-existing data on your database, you have to introspect it. Learn more in our docs:
${c.underline(c.blue("https://pris.ly/nuxt/dbpull"))}.
`
  }
};

function checkIfPrismaSchemaExists(paths) {
  const exists = paths.reduce((prev, current) => {
    return existsSync(current) || prev;
  }, false);
  if (exists) {
    consola.success(PREDEFINED_LOG_MESSAGES.checkIfPrismaSchemaExists.yes);
    return true;
  }
  consola.info(PREDEFINED_LOG_MESSAGES.checkIfPrismaSchemaExists.no);
  return false;
}
function moveEnvFileContent(dirA, dirB) {
  if (dirA === dirB) {
    return;
  }
  const envFileA = join(dirA, ".env");
  const envFileB = join(dirB, ".env");
  try {
    if (!existsSync(envFileB)) {
      consola.warn(`Source .env file does not exist in directory: ${dirB}`);
      return;
    }
    const envContentB = readFileSync(envFileB, "utf8");
    if (existsSync(envFileA)) {
      const envContentA = readFileSync(envFileA, "utf8");
      if (envContentA.match(/^DATABASE_URL=.+$/gm)) {
        return;
      }
      const combinedContent = `${envContentA}
${envContentB}`;
      writeFileSync(envFileA, combinedContent, "utf8");
    } else {
      writeFileSync(envFileA, envContentB, "utf8");
    }
    consola.success(
      `Successfully moved content from ${envFileB} to ${envFileA}`
    );
  } catch (error) {
    consola.error(`Failed to move .env file content: ${error}`);
  }
}
async function initPrisma({
  directory,
  rootDir,
  provider = "sqlite",
  datasourceUrl
}) {
  const commandArgs = ["prisma", "init", "--datasource-provider"];
  commandArgs.push(provider);
  if (datasourceUrl) {
    commandArgs.push("--url");
    commandArgs.push(datasourceUrl);
  }
  consola.info(PREDEFINED_LOG_MESSAGES.initPrisma.action);
  const { stderr, exitCode } = await x("npx", commandArgs, {
    nodeOptions: {
      cwd: directory
    }
  });
  if (exitCode !== 0) {
    consola.error(PREDEFINED_LOG_MESSAGES.initPrisma.error, stderr);
    return false;
  }
  consola.success(PREDEFINED_LOG_MESSAGES.initPrisma.success);
  try {
    moveEnvFileContent(rootDir, directory);
  } catch (error) {
    consola.error(error);
  }
  return true;
}
function checkIfMigrationsFolderExists(path) {
  if (existsSync(path)) {
    consola.success(
      PREDEFINED_LOG_MESSAGES.checkIfMigrationsFolderExists.success
    );
    return true;
  }
  consola.info(PREDEFINED_LOG_MESSAGES.checkIfMigrationsFolderExists.error);
  return false;
}
async function writeToSchema(prismaSchemaPath) {
  try {
    let existingSchema = "";
    try {
      existingSchema = readFileSync(prismaSchemaPath, "utf-8");
    } catch {
      consola.error(PREDEFINED_LOG_MESSAGES.writeToSchema.errorReadingFile);
      return false;
    }
    const addModel = `model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

model Post {
  id        Int     @id @default(autoincrement())
  title     String
  content   String?
  published Boolean @default(false)
  author    User    @relation(fields: [authorId], references: [id])
  authorId  Int
}
`;
    if (existingSchema.trim().includes(addModel.trim()))
      return;
    const updatedSchema = `${existingSchema.trim()}

${addModel}`;
    writeFileSync(prismaSchemaPath, updatedSchema);
  } catch {
    consola.error(PREDEFINED_LOG_MESSAGES.writeToSchema.failedToWrite);
  }
}
async function runMigration(directory, schemaPath) {
  consola.info(PREDEFINED_LOG_MESSAGES.runMigration.action);
  const { stderr, exitCode } = await x(
    "npx",
    ["prisma", "migrate", "dev", "--name", "init"].concat(schemaPath),
    {
      nodeOptions: {
        cwd: directory
      }
    }
  );
  if (exitCode !== 0) {
    consola.error(PREDEFINED_LOG_MESSAGES.runMigration.error, stderr);
    return false;
  }
  consola.success(PREDEFINED_LOG_MESSAGES.runMigration.success);
  return true;
}
async function formatSchema(directory, schemaPath) {
  consola.info(PREDEFINED_LOG_MESSAGES.formatSchema.action);
  const { stderr, exitCode } = await x(
    "npx",
    ["prisma", "format"].concat(schemaPath),
    {
      nodeOptions: {
        cwd: directory
      }
    }
  );
  if (exitCode !== 0) {
    consola.error(PREDEFINED_LOG_MESSAGES.formatSchema.error, stderr);
    return;
  }
  consola.success(PREDEFINED_LOG_MESSAGES.formatSchema.success);
}
async function generatePrismaClient(directory, prismaSchemaPath, verboseLog = false) {
  consola.info(PREDEFINED_LOG_MESSAGES.generatePrismaClient.action);
  const { stderr, exitCode } = await x(
    "npx",
    ["prisma", "generate"].concat(prismaSchemaPath),
    {
      nodeOptions: {
        cwd: directory
      }
    }
  );
  if (exitCode !== 0) {
    consola.error(PREDEFINED_LOG_MESSAGES.generatePrismaClient.error);
    if (verboseLog) {
      consola.error(stderr);
    }
    return;
  }
  consola.success(PREDEFINED_LOG_MESSAGES.generatePrismaClient.success);
}
async function startPrismaStudio(directory, schemaLocation) {
  consola.info(PREDEFINED_LOG_MESSAGES.startPrismaStudio.action);
  try {
    const proc = x(
      "npx",
      ["prisma", "studio", "--browser", "none"].concat(schemaLocation),
      {
        nodeOptions: {
          cwd: directory
        },
        throwOnError: true
      }
    );
    proc.process?.unref();
    consola.success(PREDEFINED_LOG_MESSAGES.startPrismaStudio.success);
    consola.info(PREDEFINED_LOG_MESSAGES.startPrismaStudio.info);
    return true;
  } catch (err) {
    consola.error(PREDEFINED_LOG_MESSAGES.startPrismaStudio.error);
  }
}
async function writeClientInLib(path) {
  const existingContent = existsSync(`${path}/lib/prisma.ts`);
  try {
    if (!existingContent) {
      const prismaClient = `import { PrismaClient } from '@prisma/client'

const prismaClientSingleton = () => {
  return new PrismaClient()
}

declare const globalThis: {
  prismaGlobal: ReturnType<typeof prismaClientSingleton>;
} & typeof global;

const prisma = globalThis.prismaGlobal ?? prismaClientSingleton()

export default prisma

if (process.env.NODE_ENV !== 'production') globalThis.prismaGlobal = prisma
`;
      if (!existsSync(`${path}/lib`)) {
        mkdirSync(`${path}/lib`);
      }
      if (existsSync(`${path}/lib/prisma.ts`)) {
        consola.info(PREDEFINED_LOG_MESSAGES.writeClientInLib.found);
        return;
      }
      writeFileSync(`${path}/lib/prisma.ts`, prismaClient);
      consola.success(PREDEFINED_LOG_MESSAGES.writeClientInLib.success);
    }
  } catch (err) {
    consola.error(err);
  }
}

async function executeRequiredPrompts({
  promptForMigrate = true
}) {
  const options = [];
  if (promptForMigrate) {
    options.push({
      message: "Do you want to migrate database changes to your database?",
      options: {
        type: "confirm",
        initial: true
      },
      key: "promptForPrismaMigrate" /* PROMPT_MIGRATE */
    });
  }
  if (options.length === 0) {
    return null;
  }
  try {
    const answers = {
      ["promptForPrismaMigrate" /* PROMPT_MIGRATE */]: false
    };
    for (const { message, options: promptOptions, key } of options) {
      answers[key] = await consola$1.prompt(message, promptOptions);
    }
    return answers;
  } catch (error) {
    consola$1.error("Error during prompts execution:", error);
    return null;
  }
}

const module = defineNuxtModule({
  meta: {
    name: "@prisma/nuxt",
    configKey: "prisma"
  },
  // Default configuration options for the module
  defaults: {
    datasources: {
      db: {
        url: process.env.DATABASE_URL
        // Security: Ensure DATABASE_URL is correctly set and secure
      }
    },
    log: [],
    errorFormat: "pretty",
    writeToSchema: true,
    formatSchema: true,
    runMigration: true,
    generateClient: true,
    installStudio: true,
    autoSetupPrisma: false,
    skipPrompts: false,
    prismaRoot: void 0,
    prismaSchemaPath: void 0
  },
  async setup(options, nuxt) {
    const { resolve: resolveProject } = createResolver(nuxt.options.rootDir);
    const { resolve: resolver } = createResolver(import.meta.url);
    const runtimeDir = fileURLToPath(new URL("./runtime", import.meta.url));
    const npmLifecycleEvent = process.env?.npm_lifecycle_event;
    const skipAllPrompts = options.skipPrompts || npmLifecycleEvent === "dev:build";
    const PRISMA_SCHEMA_CMD = options.prismaSchemaPath ? ["--schema", options.prismaSchemaPath] : [];
    const prepareModule = () => {
      nuxt.options.experimental.componentIslands ||= {};
      nuxt.options.experimental.componentIslands = true;
      addPlugin(resolver("./runtime/plugin"));
      addImportsDir(resolver(runtimeDir, "composables"));
      addServerImportsDir(resolver(runtimeDir, "utils"));
      nuxt.options.vite.optimizeDeps = defu(
        nuxt.options.vite.optimizeDeps || {},
        {
          include: ["@prisma/nuxt > @prisma/client"]
        }
      );
    };
    const forceSkipPrismaSetup = import.meta.env?.SKIP_PRISMA_SETUP ?? process.env?.SKIP_PRISMA_SETUP ?? false;
    nuxt.options.runtimeConfig.public.prisma = defu(
      nuxt.options.runtimeConfig.public.prisma || {},
      {
        log: options.log,
        errorFormat: options.errorFormat
      }
    );
    if (forceSkipPrismaSetup || npmLifecycleEvent === "postinstall") {
      if (npmLifecycleEvent !== "postinstall") {
        consola.warn(PREDEFINED_LOG_MESSAGES.PRISMA_SETUP_SKIPPED_WARNING);
      }
      prepareModule();
      return;
    }
    const PROJECT_PATH = resolveProject();
    const LAYER_PATH = options.prismaRoot ? resolveProject(options.prismaRoot) : PROJECT_PATH;
    const prismaSchemaExists = checkIfPrismaSchemaExists([
      resolveProject(LAYER_PATH, "prisma", "schema.prisma"),
      resolveProject(LAYER_PATH, "prisma", "schema")
    ]);
    const prismaMigrateWorkflow = async () => {
      const migrationFolderExists = checkIfMigrationsFolderExists(
        resolveProject(LAYER_PATH, "prisma", "migrations")
      );
      if (migrationFolderExists || !options.runMigration) {
        consola.info(PREDEFINED_LOG_MESSAGES.skipMigrations);
        return;
      }
      const migrateAndFormatSchema = async () => {
        await runMigration(PROJECT_PATH, PRISMA_SCHEMA_CMD);
        if (options.formatSchema) {
          await formatSchema(PROJECT_PATH, PRISMA_SCHEMA_CMD);
        }
      };
      if (options.autoSetupPrisma && options.runMigration) {
        await migrateAndFormatSchema();
        return;
      }
      const promptResult = await executeRequiredPrompts({
        promptForMigrate: !skipAllPrompts
      });
      if (promptResult?.promptForPrismaMigrate && options.runMigration) {
        await migrateAndFormatSchema();
      }
    };
    const prismaInitWorkflow = async () => {
      await initPrisma({
        directory: LAYER_PATH,
        rootDir: PROJECT_PATH,
        provider: "sqlite"
      });
      await writeToSchema(`${LAYER_PATH}/prisma/schema.prisma`);
    };
    const prismaStudioWorkflow = async () => {
      if (!options.installStudio || npmLifecycleEvent !== "dev") {
        consola.info(PREDEFINED_LOG_MESSAGES.skipInstallingPrismaStudio);
        return;
      }
      await startPrismaStudio(PROJECT_PATH, PRISMA_SCHEMA_CMD);
      addCustomTab({
        name: "nuxt-prisma",
        title: "Prisma Studio",
        icon: "simple-icons:prisma",
        category: "server",
        view: {
          type: "iframe",
          src: "http://localhost:5555/",
          persistent: true
        }
      });
    };
    if (!prismaSchemaExists) {
      await prismaInitWorkflow();
    }
    await prismaMigrateWorkflow();
    await writeClientInLib(LAYER_PATH);
    if (options.generateClient) {
      await generatePrismaClient(
        PROJECT_PATH,
        PRISMA_SCHEMA_CMD,
        options.log?.includes("error")
      );
    }
    await prismaStudioWorkflow();
    prepareModule();
  }
});

export { module as default };
