// @nuxtjs/mcp-toolkit/dist/runtime/server/mcp/definitions/tools.js
import { enrichNameTitle } from "./utils.js";
import { createCacheOptions, wrapWithCache } from "./cache.js";
export function registerToolFromDefinition(server, tool) {
  const { name, title } = enrichNameTitle({
    name: tool.name,
    title: tool.title,
    _meta: tool._meta,
    type: "tool"
  });
  let handler = tool.handler;
  if (tool.cache !== void 0) {
    const defaultGetKey = tool.inputSchema ? (args) => {
      const values = Object.values(args);
      return values.map((v) => String(v).replace(/\//g, "-").replace(/^-/, "")).join(":");
    } : void 0;
    const cacheOptions = createCacheOptions(tool.cache, `mcp-tool:${name}`, defaultGetKey);
    handler = wrapWithCache(
      tool.handler,
      cacheOptions
    );
  }
  const options = {
    title,
    description: tool.description,
    inputSchema: tool.inputSchema,
    outputSchema: tool.outputSchema,
    annotations: tool.annotations,
    _meta: {
      ...tool._meta,
      ...tool.inputExamples && { inputExamples: tool.inputExamples }
    }
  };
  return server.registerTool(name, options, handler);
}
export function defineMcpTool(definition) {
  return definition;
}
