// @nuxtjs/html-validator/dist/module.mjs
import { fileURLToPath, pathToFileURL } from 'node:url';
import { colors } from 'consola/utils';
import { normalize } from 'pathe';
import { isWindows } from 'std-env';
import { genObjectFromRawEntries, genArrayFromRaw } from 'knitwork';
import { defineNuxtModule, createResolver, logger, resolvePath } from '@nuxt/kit';

const defaultHtmlValidateConfig = {
  extends: [
    "html-validate:document",
    "html-validate:recommended",
    "html-validate:standard"
  ],
  rules: {
    //
    "svg-focusable": "off",
    "no-unknown-elements": "error",
    // Conflicts or not needed as we use prettier formatting
    "void-style": "off",
    "no-trailing-whitespace": "off",
    // Conflict with Nuxt defaults
    "require-sri": "off",
    "attribute-boolean-style": "off",
    "doctype-style": "off",
    // Unreasonable rule
    "no-inline-style": "off"
  }
};
const DEFAULTS = {
  usePrettier: false,
  enabled: true,
  failOnError: false,
  options: defaultHtmlValidateConfig,
  hookable: false,
  ignore: [/\.(xml|rss|json)$/]
};
const NuxtRedirectHtmlRegex = /<!DOCTYPE html><html><head><meta http-equiv="refresh" content="0; url=([^"]+)"><\/head><\/html>/;

const module = defineNuxtModule({
  meta: {
    name: "@nuxtjs/html-validator",
    configKey: "htmlValidator",
    compatibility: {
      nuxt: ">=3.0.0-rc.7"
    }
  },
  defaults: (nuxt) => ({
    ...DEFAULTS,
    logLevel: nuxt.options.dev ? "verbose" : "warning"
  }),
  async setup(moduleOptions, nuxt) {
    const resolver = createResolver(
      import.meta.url
    );
    nuxt.hook("prepare:types", ({ references }) => {
      const types = resolver.resolve("./runtime/types.d.ts");
      references.push({ path: types });
    });
    if (nuxt.options._prepare || moduleOptions.enabled === false) {
      return;
    }
    logger.info(`Using ${colors.bold("html-validate")} to validate server-rendered HTML`);
    const { usePrettier, failOnError, options, logLevel } = moduleOptions;
    if (nuxt.options.htmlValidator?.options?.extends) {
      options.extends = nuxt.options.htmlValidator.options.extends;
    }
    if (nuxt.options.dev) {
      nuxt.hook("nitro:config", (config) => {
        config.externals = config.externals || {};
        config.externals.inline = config.externals.inline || [];
        config.externals.inline.push("@nuxtjs/html-validator");
        config.plugins = config.plugins || [];
        config.plugins.push(normalize(fileURLToPath(new URL("./runtime/nitro", import.meta.url))));
        config.virtual = config.virtual || {};
        const serialisedOptions = genObjectFromRawEntries(Object.entries(moduleOptions).map(([key, value]) => {
          if (key !== "ignore") return [key, JSON.stringify(value, null, 2)];
          const ignore = value || [];
          return [key, genArrayFromRaw(ignore.map((v) => typeof v === "string" ? JSON.stringify(v) : v.toString()))];
        }));
        config.virtual["#html-validator-config"] = `export default ${serialisedOptions}`;
      });
    }
    if (!nuxt.options.dev) {
      const validatorPath = await resolvePath(fileURLToPath(new URL("./runtime/validator", import.meta.url)));
      const { useChecker, getValidator } = await (isWindows ? import(pathToFileURL(validatorPath).href) : import(validatorPath));
      const validator = getValidator(options);
      const { checkHTML, invalidPages } = useChecker(validator, usePrettier, logLevel);
      if (failOnError) {
        const errorIfNeeded = () => {
          if (invalidPages.length) {
            throw new Error("html-validator found errors");
          }
        };
        nuxt.hook("close", errorIfNeeded);
      }
      nuxt.hook("nitro:init", (nitro) => {
        nitro.hooks.hook("prerender:generate", (route) => {
          if (!route.contents || !route.fileName?.endsWith(".html")) {
            return;
          }
          if (route.contents.match(NuxtRedirectHtmlRegex)) {
            return;
          }
          checkHTML(route.route, route.contents);
        });
      });
    }
  }
});

export { module as default };
