// @nuxtjs/mcp-toolkit/dist/module.mjs
import { logger, createResolver, defineNuxtModule, addComponent, addServerTemplate, addServerHandler, addServerImports } from '@nuxt/kit';
import { loadAllDefinitions } from '../dist/runtime/server/mcp/loaders/index.js';
import { defaultMcpConfig, getMcpConfig } from '../dist/runtime/server/mcp/config.js';
import { ROUTES } from '../dist/runtime/server/mcp/constants.js';
import { addDevToolsCustomTabs } from '../dist/runtime/server/mcp/devtools/index.js';
import { execSync } from 'node:child_process';
import { existsSync, readFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

const IDE_CONFIGS = {
  cursor: { name: "Cursor" },
  vscode: { name: "VS Code" }
};
function terminalLink(text, url) {
  return `\x1B]8;;${url}\x07${text}\x1B]8;;\x07`;
}
function isMCPInstalled(mcpJson, mcpUrl) {
  if (!mcpJson.mcpServers) return false;
  return Object.values(mcpJson.mcpServers).some((server) => server.url === mcpUrl);
}
function detectIDE() {
  const env = process.env;
  if (env.__CFBundleIdentifier === "com.todesktop.230313mzl4w4u92") return "cursor";
  if (env.__CFBundleIdentifier === "com.microsoft.VSCode") return "vscode";
  if (env.CURSOR_TRACE_ID) return "cursor";
  const ipc = env.VSCODE_IPC_HOOK || "";
  if (ipc.includes("/Cursor/")) return "cursor";
  if (ipc.includes("/Code/")) return "vscode";
  try {
    let pid = process.ppid;
    for (let i = 0; i < 10 && pid > 1; i++) {
      const name = execSync(`ps -o comm= -p ${pid}`, { stdio: ["pipe", "pipe", "ignore"] }).toString().toLowerCase();
      if (name.includes("cursor")) return "cursor";
      if (name.includes("code helper") || name.includes("code.app")) return "vscode";
      pid = Number.parseInt(execSync(`ps -o ppid= -p ${pid}`, { stdio: ["pipe", "pipe", "ignore"] }).toString().trim());
    }
  } catch {
  }
  return null;
}
function getMCPConfigPaths(ide, rootDir) {
  return [
    {
      path: join(rootDir, `.${ide}/mcp.json`),
      displayPath: `.${ide}/mcp.json`
    },
    {
      path: join(homedir(), `.${ide}/mcp.json`),
      displayPath: `~/.${ide}/mcp.json`
    }
  ];
}
function findInstalledMCPConfig(ide, rootDir, mcpUrl) {
  const configPaths = getMCPConfigPaths(ide, rootDir);
  for (const config of configPaths) {
    if (existsSync(config.path)) {
      try {
        const mcpJson = JSON.parse(readFileSync(config.path, "utf8"));
        if (isMCPInstalled(mcpJson, mcpUrl)) {
          return config;
        }
      } catch {
      }
    }
  }
  return null;
}
function generateDeeplinkUrl(baseUrl, route, ide, serverName) {
  const mcpName = `local-${serverName.toLowerCase().replace(/\s+/g, "-")}`;
  return `${baseUrl}${route}/deeplink?ide=${ide}&name=${encodeURIComponent(mcpName)}`;
}

const name = "@nuxtjs/mcp-toolkit";
const version = "0.7.0";

const log = logger.withTag("@nuxtjs/mcp-toolkit");
const { resolve } = createResolver(import.meta.url);
const module$1 = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "mcp",
    docs: "https://mcp-toolkit.nuxt.dev/getting-started/installation",
    mcp: "https://mcp-toolkit.nuxt.dev/mcp"
  },
  defaults: defaultMcpConfig,
  async setup(options, nuxt) {
    if (nuxt.options.nitro.static || nuxt.options._generate) {
      log.warn("@nuxtjs/mcp-toolkit is not compatible with `nuxt generate` as it needs a server to run.");
      return;
    }
    const resolver = createResolver(import.meta.url);
    const mcpConfig = getMcpConfig(options);
    if (!options.enabled) {
      return;
    }
    addComponent({
      name: "InstallButton",
      filePath: resolver.resolve("runtime/components/InstallButton.vue")
    });
    addServerTemplate({
      filename: "#nuxt-mcp-toolkit/config.mjs",
      getContents: () => `export default ${JSON.stringify(mcpConfig)}`
    });
    const mcpDir = mcpConfig.dir;
    const paths = {
      tools: [`${mcpDir}/tools`],
      resources: [`${mcpDir}/resources`],
      prompts: [`${mcpDir}/prompts`],
      handlers: [mcpDir]
    };
    let mcpSummary = null;
    nuxt.hook("modules:done", async () => {
      try {
        const callCustomHook = nuxt.callHook;
        await callCustomHook("mcp:definitions:paths", paths);
        const result = await loadAllDefinitions(paths);
        if (result.handlers && result.handlers.count > 0) {
          addServerHandler({
            route: ROUTES.CUSTOM_HANDLER,
            handler: resolver.resolve("runtime/server/mcp/handler")
          });
        }
        if (result.total === 0) {
          log.warn("No MCP definitions found. Create tools, resources, or prompts in server/mcp/");
        } else {
          const summary = [];
          if (result.tools.count > 0) summary.push(`${result.tools.count} tool${result.tools.count > 1 ? "s" : ""}`);
          if (result.resources.count > 0) summary.push(`${result.resources.count} resource${result.resources.count > 1 ? "s" : ""}`);
          if (result.prompts.count > 0) summary.push(`${result.prompts.count} prompt${result.prompts.count > 1 ? "s" : ""}`);
          if (result.handlers.count > 0) summary.push(`${result.handlers.count} handler${result.handlers.count > 1 ? "s" : ""}`);
          mcpSummary = summary.join(", ");
        }
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        log.error("Failed to initialize MCP server");
        log.error(`Error: ${errorMessage}`);
        throw error;
      }
    });
    nuxt.hook("listen", (_, listener) => {
      if (!mcpSummary) return;
      const ide = detectIDE();
      if (ide) {
        const baseUrl = listener.url.replace(/\/$/, "");
        const mcpUrl = `${baseUrl}${options.route}`;
        const installedConfig = findInstalledMCPConfig(ide, nuxt.options.rootDir, mcpUrl);
        if (installedConfig) {
          log.success(`\`${options.route}\` enabled with ${mcpSummary} \xB7 MCP server already installed in \`${installedConfig.displayPath}\``);
          return;
        }
        const ideName = IDE_CONFIGS[ide].name;
        const deeplinkUrl = generateDeeplinkUrl(baseUrl, options.route, ide, options.name || "mcp-server");
        log.info(`${ideName} detected. ${terminalLink("Install Nuxt MCP server", deeplinkUrl)}`);
        log.success(`\`${options.route}\` enabled with ${mcpSummary}`);
      } else {
        log.success(`\`${options.route}\` enabled with ${mcpSummary}`);
      }
    });
    nuxt.hook("prepare:types", ({ references }) => {
      references.push({
        path: resolver.resolve("runtime/server/types.server.d.ts")
      });
    });
    nuxt.options.nitro.typescript ??= {};
    nuxt.options.nitro.typescript.tsConfig ??= {};
    nuxt.options.nitro.typescript.tsConfig.include ??= [];
    nuxt.options.nitro.typescript.tsConfig.include.push(resolver.resolve("runtime/server/types.server.d.ts"));
    let isCloudflare = false;
    if (!nuxt.options.dev) {
      nuxt.hook("nitro:config", (nitroConfig) => {
        const preset = nitroConfig.preset || process.env.NITRO_PRESET || "";
        isCloudflare = preset.includes("cloudflare");
      });
    }
    addServerTemplate({
      filename: "#nuxt-mcp-toolkit/transport.mjs",
      getContents: () => {
        const provider = isCloudflare ? "cloudflare" : "node";
        return `export { default } from '${resolver.resolve(`runtime/server/mcp/providers/${provider}`)}'`;
      }
    });
    const mcpDefinitionsPath = resolver.resolve("runtime/server/mcp/definitions");
    addServerImports([
      "defineMcpTool",
      "defineMcpResource",
      "defineMcpPrompt",
      "defineMcpHandler",
      "textResult",
      "jsonResult",
      "errorResult",
      "imageResult"
    ].map((name2) => ({ name: name2, from: mcpDefinitionsPath })));
    addServerHandler({
      route: options.route,
      handler: resolver.resolve("runtime/server/mcp/handler")
    });
    addServerHandler({
      route: `${options.route}/deeplink`,
      handler: resolver.resolve("runtime/server/mcp/deeplink")
    });
    addServerHandler({
      route: `${options.route}/badge.svg`,
      handler: resolver.resolve("runtime/server/mcp/badge-image")
    });
    addDevToolsCustomTabs(nuxt, options);
  }
});

export { module$1 as default, resolve };
